File: server/src/game/Hero.ts
import { MAX_HERO_HEALTH } from './constants.js';

// Hero của một player — health được quản lý bên trong class (encapsulation).
export class Hero {
  private health: number;

  constructor(
    public readonly heroId: string,
    public readonly name: string,
    public readonly heroClass: string,
    public readonly powerName: string,
    public readonly powerCost: number,
    public readonly imagePath: string,
    health: number = MAX_HERO_HEALTH,
  ) {
    this.health = health;
  }

  get currentHealth(): number {
    return this.health;
  }

  get maxHealth(): number {
    return MAX_HERO_HEALTH;
  }

  takeDamage(amount: number): number {
    if (!Number.isFinite(amount) || amount < 0) throw new Error('Damage không hợp lệ.');
    const actual = Math.min(Math.max(0, this.health), amount);
    this.health -= actual;
    return actual;
  }

  heal(amount: number): void {
    if (!Number.isFinite(amount) || amount < 0) throw new Error('Heal không hợp lệ.');
    this.health = Math.min(this.maxHealth, this.health + amount);
  }

  isDead(): boolean {
    return this.health <= 0;
  }

  checkpoint(): () => void {
    const health = this.health;
    return () => { this.health = health; };
  }
}

File: server-java/src/main/java/vn/coincard/server/game/Hero.java
package vn.coincard.server.game;

import vn.coincard.server.model.GameCharacter;

/** Mutable hero — Inheritance: Hero là một GameCharacter. */
public class Hero extends GameCharacter {
  public Hero(String heroId, String name, HeroClass heroClass,
      String powerName, int powerCost, String imagePath) {
    this(heroId, name, heroClass, powerName, powerCost, imagePath, Constants.MAX_HERO_HEALTH);
  }

  public Hero(String heroId, String name, HeroClass heroClass,
      String powerName, int powerCost, String imagePath, int health) {
    super(Constants.MAX_HERO_HEALTH);
    this.heroId = heroId;
    this.name = name;
    this.heroClass = heroClass;
    this.powerName = powerName;
    this.powerCost = powerCost;
    this.imagePath = imagePath;
    setCurrentHealth(Math.min(health, Constants.MAX_HERO_HEALTH));
  }

  private final String heroId;
  private final String name;
  private final HeroClass heroClass;
  private final String powerName;
  private final int powerCost;
  private final String imagePath;

  public String getHeroId() { return heroId; }
  public String getName() { return name; }
  public HeroClass getHeroClass() { return heroClass; }
  public String getPowerName() { return powerName; }
  public int getPowerCost() { return powerCost; }
  public String getImagePath() { return imagePath; }

  @Override
  public int takeDamage(int amount) {
    int before = currentHealth();
    super.takeDamage(amount);
    return before - currentHealth();
  }

  record State(int currentHealth, int maxHealth) {}

  State snapshotState() {
    return new State(currentHealth(), maxHealth());
  }

  void restoreState(State state) {
    setMaxHealth(state.maxHealth());
    setCurrentHealth(state.currentHealth());
  }
}
